### ISSUE TITLE
Mouse keys

### ISSUE THREAD

**Issue by oleic** [kasper/phoenix#90]
Hey,

is it possible to get mouse keys available?

**Comment by kasper** [comment]
Hello! If you mean that you would like to use function buttons on your mouse to trigger Phoenix, you can usually use the driver software for your mouse to rebind those buttons to a custom keyboard shortcut. Then you can bind that shortcut with Phoenix. Would that help?

**Comment by oleic** [comment]
I actually want to check in javascript if the mouse button is up or down. Is there a way to check that?

**Comment by kasper** [comment]
Ah, I see. There’s currently no way to determine that, but I presume it would be possible. Could you describe how you would use this?

**Comment by oleic** [comment]
I am trying to implement a window snap feature and like to know, if the
user has released the mouse button or not. The window should only snap at
the edge, if the mouse button is released.
Kasper Hirvikoski <notifications@github.com> schrieb am Fr., 6. Mai 2016 um
08:30:

**Comment by kasper** [comment]
Right, so you’re presumably listening to the `windowDidMove`-event? For now, you could use something like the [debounce](http://underscorejs.org/#debounce)-function from Underscore to determine when an event has stopped occuring. I think that would be the easiest way to implement a snapping feature…

I’ll think about how we could support mouse events in the future.

**Comment by oleic** [comment]
Sounds great! Thanks for the tip using debounce. :)

2016-05-06 10:57 GMT+02:00 Kasper Hirvikoski <notifications@github.com>:

**Comment by kasper** [comment]
You’re most welcome. Let me know if you need more help with debounce.

**Comment by mafredri** [comment]
I noticed `setTimeout` is missing from Phoenix, so I tried, out of interest, to use the underscore debounce (since it uses `setTimeout`) and noticed that it doesn't work. Phoenix prints out a `ReferenceError`. Is it supposed to work, am I missing something?

**Comment by kasper** [comment]
@mafredri Darn, you’re right! I completely forgot it obviously uses `setTimeout`. JavaScriptCore doesn’t include timing functions, in browsers those are implemented in the DOM.

Nevertheless, Phoenix does support timing functions, check out `Phoenix.after` and `Phoenix.every`. I guess you would need to first implement `setTimeout` with them.

**Comment by mafredri** [comment]
I took a stab at implementing the timeout/interval API for Phoenix. It also fixes the underscore debounce functionality :). Maybe of interest to you @oleic?

I tried to keep it simple, whilst removing the explicit requirement on keeping references to your handles (when using setTimeout / setInterval).

```javascript
(function(global) {
	// Keeps references to Phoenix handlers
	var refs = new Set();

	// Add functions on the global object
	global.setTimeout = setTimeout;
	global.clearTimeout = clear;
	global.setInterval = setInterval;
	global.clearInterval = clear;

	function setTimeout(fn, interval) {
		var handler = Phoenix.after(interval / 1000, handleFn);
		refs.add(handler);
		return handler;

		function handleFn() {
			fn();
			refs.delete(handler); // Delete reference after completion
		}
	}

	function setInterval(fn, interval) {
		var handler = Phoenix.every(interval / 1000, fn);
		refs.add(handler);
		return handler;
	}

	function clear(handler) {
		handler.stop(); // Stop the timer immediately
		refs.delete(handler);
	}
})(this);
```

**Comment by kasper** [comment]
@mafredri Nice! I think we should add `setTimeout` and `setInterval` to the core to enable various timing related functions from Underscore.

**Comment by mafredri** [comment]
I agree, it would make sense to support all functionality of Underscore since it comes bundled! And also, supporting commonly used APIs makes it more fluent to use Phoenix :).

**Comment by kasper** [comment]
I’ve added support for the following mouse events to Phoenix. Others can be added provided they are useful. Interestingly, there doesn’t seem to be any distinction between “down” and “up” with global mouse events. Both are triggered when the click is released — so I’ve only provided the up events. This feature will be released in 2.2.

- `mouseDidMove` triggered when the mouse has moved
- `mouseDidLeftClick` triggered when the mouse did left click
- `mouseDidRightClick` triggered when the mouse did right click
